src/main/java/com/marcusslover/plus/lib/sound/Music.java

```diff
@@ -0,0 +1,223 @@
+        return new Music(intro, introLength, loop, loopLength, tail);
+    }
+
+    /**
+     * Stop music from playing to all audiences.
+     */
+    public static void stopAll() {
+        for (Audience audience : Music.playerMusic.keySet()) {
+            Music.playerMusic.get(audience).forceStop(audience);
+        }
+    }
+
+    /**
+     * Stop the music for an audience and play the tail if it exists.
+     *
+     * @param audience The audience.
+     */
+    public void stop(Audience audience) {
+        if (!Music.playerMusic.containsKey(audience)) {
+            return;
+        }
+        stopTasks();
+        if (this.loops > 0) {
+            Task.syncDelayed(ServerUtils.getCallingPlugin(), () -> {
+                audience.stopSound(this.loop.sound);
+                if (this.tail != null) {
+                    this.tail.send(audience);
+                }
+            }, this.loopLength - (this.ticks % this.loopLength));
+        } else {
+            audience.stopSound(this.loop.sound);
+            if (this.tail != null) {
+                this.tail.send(audience);
+            }
+        }
+        Music.playerMusic.remove(audience);
+    }
+
+    /**
+     * Stop the music from the audience without playing the tail.
+     *
+     * @param audience The audience.
+     */
+    public void forceStop(Audience audience) {
+        if (!Music.playerMusic.containsKey(audience)) {
+            return;
+        }
+        stopTasks();
+        if (this.intro != null) {
+            audience.stopSound(this.intro().sound);
+        }
+        if (this.loop != null) {
+            audience.stopSound(this.loop().sound);
+        }
+        if (this.tail != null) {
+            audience.stopSound(this.tail().sound);
+        }
+        Music.playerMusic.remove(audience);
+    }
+
+    /**
+     * Stops the music from looping again, but it will finish the current note.
+     */
+    public void stopTasks() {
+        this.tickTask.cancel();
+        this.loopTask.cancel();
+    }
+
+    @Override
+    public @NotNull <T extends CommandSender> Music send(@NotNull T target) throws IllegalStateException {
+        if (Music.playerMusic.containsKey(target)) {
+            throw new IllegalStateException("Music is already playing for this target.");
+        } else {
+            Music.playerMusic.put(target, this);
+        }
+        if (this.introLength != 0 && this.intro != null) {
+            this.intro.send(target);
+            loopTask = Task.syncRepeating(ServerUtils.getCallingPlugin(), () -> {
+                this.loop.send(target);
+                this.loops++;
+            }, this.introLength + 20, loopLength);
+        } else {
+            loopTask = Task.syncRepeating(ServerUtils.getCallingPlugin(), () -> {
+                this.loop.send(target);
+                this.loops++;
+            }, 1, loopLength);
+        }
```

Comment:
Is there also any way that this would be able to detect client volume changes, because currently, when you turn the channel that you send the music to to 0% and back up, it doesn't play until the loop replay's it. I am working on the session now.

Reply:
@joshuaepstein I believe that is client sided. No way to detect if a player changes their client settings.